NEURON 9.0 test failures

Hello!

We've recently released NEURON version 9.0.0: https://pypi.org/project/neuron/9.0.0/

I see that this release seems to have affected the CI of netpyne: https://github.com/suny-downstate-medical-center/netpyne/actions/runs/18328606470/job/52198894276

Pasting the relevant info from the log:

```plaintext
=========================== short test summary info ============================
FAILED tests/doc/test_tutorials.py::test_tutorial_1 - AssertionError
FAILED tests/doc/test_tutorials.py::test_tutorial_2 - AssertionError
FAILED tests/doc/test_tutorials.py::test_tutorial_3 - AssertionError
FAILED tests/doc/test_tutorials.py::test_tutorial_5 - AssertionError
FAILED tests/doc/test_tutorials.py::test_tutorial_6 - AssertionError
FAILED tests/doc/test_tutorials.py::test_tutorial_7 - AssertionError
FAILED tests/doc/test_tutorials.py::test_tutorial_8 - assert 13.75 == 13.675
======================== 7 failed, 1 warning in 29.53s =========================
```

We are also experiencing the exact same issue as part of the EBRAINS 2.0 project, so it'd be good to narrow down the source. Could you elaborate a bit on what these tests do (i.e. what Python/hoc script they run), and what mod files do they use?

For reference, the following two NEURON issues do not seem to be related to the above (first things that came to mind):

- https://github.com/neuronsimulator/nrn/issues/1764 (I attempted to patch netpyne to use the previous results since we put back the behavior of NEURON 8.0 in 9.0, but that didn't work)
- https://github.com/neuronsimulator/nrn/issues/3137#issuecomment-3044581057 (reverting the linked PR did not change the results above)

---

To reproduce locally (tested on Ubuntu 24.04 LTS x86_64), I did the following:

- `git clone https://github.com/suny-downstate-medical-center/netpyne/`
- `cd netpyne`
- `python3 -m venv env && source env/bin/activate`
- `pip install uv && uv pip install -r requirements.txt 'neuron>=9' .`
- `bash test/tests.sh`

--- a/examples/NeuroMLImport/poissonFiringSyn.mod
+++ b/examples/NeuroMLImport/poissonFiringSyn.mod
@@ -209,8 +209,13 @@ PROCEDURE seed(x) {
 }
 
 VERBATIM
+#ifndef NRN_VERSION_GTEQ_8_2_0
 double nrn_random_pick(void* r);
 void* nrn_random_arg(int argpos);
+#define RANDCAST
+#else
+#define RANDCAST (Rand*)
+#endif
 ENDVERBATIM
 
 
@@ -222,7 +227,7 @@ VERBATIM
 		: each instance. However, the corresponding hoc Random
 		: distribution MUST be set to Random.negexp(1)
 		*/
-		_lerand = nrn_random_pick(_p_donotuse);
+		_lerand = nrn_random_pick(RANDCAST _p_donotuse);
 	}else{
 		/* only can be used in main thread */
 		if (_nt != nrn_threads) {

--- a/examples/PTcell/mod/vecstim.mod
+++ b/examples/PTcell/mod/vecstim.mod
@@ -41,9 +41,11 @@ NET_RECEIVE (w) {
 }
 
 VERBATIM
+#ifndef NRN_VERSION_GTEQ_8_2_0
 extern double* vector_vec(void *);
 extern int vector_capacity(void *);
 extern void* vector_arg(int);
+#endif
 ENDVERBATIM
 
 PROCEDURE element() {

--- a/examples/RL_arm/nsloc.mod
+++ b/examples/RL_arm/nsloc.mod
@@ -79,8 +79,13 @@ FUNCTION invl(mean (ms)) (ms) {
 	}
 }
 VERBATIM
+#ifndef NRN_VERSION_GTEQ_8_2_0
 double nrn_random_pick(void* r);
 void* nrn_random_arg(int argpos);
+#define RANDCAST
+#else
+#define RANDCAST (Rand*)
+#endif
 ENDVERBATIM
 
 FUNCTION erand() {
@@ -91,7 +96,7 @@ VERBATIM
 		: each instance. However, the corresponding hoc Random
 		: distribution MUST be set to Random.negexp(1)
 		*/
-		_lerand = nrn_random_pick(_p_donotuse);
+		_lerand = nrn_random_pick(RANDCAST _p_donotuse);
 	}else{
 		/* only can be used in main thread */
 		if (_nt != nrn_threads) {

--- a/examples/batch_rxd_net/mod/vecstim.mod
+++ b/examples/batch_rxd_net/mod/vecstim.mod
@@ -41,9 +41,11 @@ NET_RECEIVE (w) {
 }
 
 VERBATIM
+#ifndef NRN_VERSION_GTEQ_8_2_0
 extern double* vector_vec(void *);
 extern int vector_capacity(void *);
 extern void* vector_arg(int);
+#endif
 ENDVERBATIM
 
 PROCEDURE element() {

--- a/examples/rxd_net/mod/vecstim.mod
+++ b/examples/rxd_net/mod/vecstim.mod
@@ -41,9 +41,11 @@ NET_RECEIVE (w) {
 }
 
 VERBATIM
+#ifndef NRN_VERSION_GTEQ_8_2_0
 extern double* vector_vec(void *);
 extern int vector_capacity(void *);
 extern void* vector_arg(int);
+#endif
 ENDVERBATIM
 
 PROCEDURE element() {

--- a/netpyne/batchtools/examples/wscale/mod/ProbAMPANMDA_EMS.mod
+++ b/netpyne/batchtools/examples/wscale/mod/ProbAMPANMDA_EMS.mod
@@ -107,8 +107,13 @@ VERBATIM
 #include<stdio.h>
 #include<math.h>
 
+#ifndef NRN_VERSION_GTEQ_8_2_0
 double nrn_random_pick(void* r);
 void* nrn_random_arg(int argpos);
+#define RANDCAST
+#else
+#define RANDCAST (Rand*)
+#endif
 
 ENDVERBATIM
   
@@ -308,7 +313,7 @@ VERBATIM
                 : each instance. However, the corresponding hoc Random
                 : distribution MUST be set to Random.negexp(1)
                 */
-                value = nrn_random_pick(_p_rng);
+                value = nrn_random_pick(RANDCAST _p_rng);
                 //printf("random stream for this simulation = %lf\n",value);
                 return value;
         }else{

--- a/netpyne/batchtools/examples/wscale/mod/ProbGABAAB_EMS.mod
+++ b/netpyne/batchtools/examples/wscale/mod/ProbGABAAB_EMS.mod
@@ -105,8 +105,13 @@ VERBATIM
 #include<stdio.h>
 #include<math.h>
 
+#ifndef NRN_VERSION_GTEQ_8_2_0
 double nrn_random_pick(void* r);
 void* nrn_random_arg(int argpos);
+#define RANDCAST
+#else
+#define RANDCAST (Rand*)
+#endif
 
 ENDVERBATIM
   
@@ -307,7 +312,7 @@ VERBATIM
                 : each instance. However, the corresponding hoc Random
                 : distribution MUST be set to Random.uniform(1)
                 */
-                value = nrn_random_pick(_p_rng);
+                value = nrn_random_pick(RANDCAST _p_rng);
                 //printf("random stream for this simulation = %lf\n",value);
                 return value;
         }else{

--- a/netpyne/batchtools/examples/wscale/mod/vecstim.mod
+++ b/netpyne/batchtools/examples/wscale/mod/vecstim.mod
@@ -41,9 +41,11 @@ NET_RECEIVE (w) {
 }
 
 VERBATIM
+#ifndef NRN_VERSION_GTEQ_8_2_0
 extern double* vector_vec(void *);
 extern int vector_capacity(void *);
 extern void* vector_arg(int);
+#endif
 ENDVERBATIM
 
 PROCEDURE element() {

--- a/netpyne/sim/setup.py
+++ b/netpyne/sim/setup.py
@@ -6,7 +6,7 @@ Module for setting up simulations
 import sys
 import os
 import numpy as np
-from neuron import h  # Import NEURON
+from neuron import h, __version__ as neuron_version # Import NEURON
 from .. import specs
 from ..specs import Dict, ODict
 from . import utils, validator
@@ -362,9 +362,10 @@ def setupRecordLFP():
                 )  # transfer resistance for each cell
 
             cell.imembPtr = h.PtrVector(nseg)  # pointer vector
-            cell.imembPtr.ptr_update_callback(
-                cell.setImembPtr
-            )  # used for gathering an array of  i_membrane values from the pointer vector
+            if neuron_version < '9.0.0':
+                cell.imembPtr.ptr_update_callback(
+                    cell.setImembPtr
+                )  # used for gathering an array of  i_membrane values from the pointer vector
             cell.imembVec = h.Vector(nseg)
 
         sim.cvode.use_fast_imem(True)  # make i_membrane_ a range variable
@@ -427,9 +428,10 @@ def setupRecordDipole():
             # set up recording of membrane currents (duplicate with setupRecordLFP -- unifiy and avoid calling twice)
             nseg = cell.getNumberOfSegments()
             cell.imembPtr = h.PtrVector(nseg)  # pointer vector
-            cell.imembPtr.ptr_update_callback(
-                cell.setImembPtr
-            )  # used for gathering an array of  i_membrane values from the pointer vector
+            if neuron_version < '9.0.0':
+                cell.imembPtr.ptr_update_callback(
+                    cell.setImembPtr
+                )  # used for gathering an array of  i_membrane values from the pointer vector
             cell.imembVec = h.Vector(nseg)
 
         sim.cvode.use_fast_imem(True)  # make i_membrane_ a range variable

--- a/netpyne/tests/checks.py
+++ b/netpyne/tests/checks.py
@@ -63,7 +63,7 @@ def checkOutput(modelName, verbose=False):
                 'cell_lfp': 1,
                 'saving': 3699,
             }
-        else:
+        elif neuron.__version__ >= '8.1.0' and neuron.__version__ < '9.0.0':
             expectedSpikes = {
                 'tut1': 2053,
                 'tut2': 930,
@@ -81,6 +81,30 @@ def checkOutput(modelName, verbose=False):
                 'cell_lfp': 1,
                 'saving': 3624,
             }
+        else:
+            # NEURON 9 has a modified netstim.mod which uses a better RNG
+            # it also reintroduces the TABLE statement in hh.mod
+            # for more details, see the following links:
+            # - https://github.com/neuronsimulator/nrn/pull/2627
+            # - https://github.com/neuronsimulator/nrn/pull/3093
+            expectedSpikes = {
+                'tut1': 2067,
+                'tut2': 933,
+                'tut3': 572,
+                'tut4': 1186,
+                'tut5': 4871,
+                'tut6': 137,
+                'tut7': 328,
+                'tut_import': 3135,
+                'HHTut': 2067,
+                'HybridTut': 2664,
+                'M1': 14572,
+                'M1detailed': 2880,
+                'PTcell': 4,
+                'cell_lfp': 1,
+                'saving': 3750,
+            }
+
         expectedAll = {'numSyns': expectedSyns, 'numSpikes': expectedSpikes}
 
         # compare all features

--- a/netpyne/tutorials/mod/vecstim.mod
+++ b/netpyne/tutorials/mod/vecstim.mod
@@ -41,9 +41,11 @@ NET_RECEIVE (w) {
 }
 
 VERBATIM
+#ifndef NRN_VERSION_GTEQ_8_2_0
 extern double* vector_vec(void *);
 extern int vector_capacity(void *);
 extern void* vector_arg(int);
+#endif
 ENDVERBATIM
 
 PROCEDURE element() {

--- a/tests/doc/test_tutorials.py
+++ b/tests/doc/test_tutorials.py
@@ -79,4 +79,4 @@ def test_tutorial_8():
     p.join()
     # check the output
     test = json.load(open("/tmp/tut8_data/tauWeight_0_0_data.json", "r"))
-    assert test["simData"]["avgRate"] == 13.675
+    assert test["simData"]["avgRate"] == 13.75

--- a/tests/doc/test_tutorials.py
+++ b/tests/doc/test_tutorials.py
@@ -73,10 +73,14 @@ def batch_example():
 
 def test_tutorial_8():
     """test tutorial 8 grid parameter search"""
+    from neuron import __version__
 
     p = Process(target=batch_example)
     p.start()
     p.join()
     # check the output
     test = json.load(open("/tmp/tut8_data/tauWeight_0_0_data.json", "r"))
-    assert test["simData"]["avgRate"] == 13.75
+    if __version__ < '9.0.0':
+        assert test["simData"]["avgRate"] == 13.675
+    else:
+        assert test["simData"]["avgRate"] == 13.75

